controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiHandler.java

```diff
@@ -328,6 +330,58 @@ private Application getApplication(String tenantName, String applicationName) {
                           .orElseThrow(() -> new NotExistsException(applicationId + " not found"));
     }
 
+    private HttpResponse nodes(String tenantName, String applicationName, String instanceName, String environment, String region) {
+        ApplicationId id = ApplicationId.from(tenantName, applicationName, instanceName);
+        ZoneId zone = ZoneId.from(environment, region);
+        List<Node> nodes = controller.configServer().nodeRepository().list(zone, id);
+
+        Slime slime = new Slime();
+        Cursor nodesArray = slime.setObject().setArray("nodes");
+        for (Node node : nodes) {
+            Cursor nodeObject = nodesArray.addObject();
+            nodeObject.setString("hostname", node.hostname().value());
+            nodeObject.setString("state", valueOf(node.state()));
+            nodeObject.setString("orchestration", valueOf(node.serviceState()));
+            nodeObject.setString("version", node.currentVersion().toString());
+            nodeObject.setString("flavor", node.canonicalFlavor());
+            nodeObject.setString("clusterId", node.clusterId());
+            nodeObject.setString("clusterType", valueOf(node.clusterType()));
+        }
+        return new SlimeJsonResponse(slime);
+    }
+
+    private static String valueOf(Node.State state) {
+        switch (state) {
+            case failed: return "failed";
+            case parked: return "parked";
+            case dirty: return "dirty";
+            case ready: return "ready";
+            case active: return "active";
+            case inactive: return "inactive";
+            case reserved: return "reserved";
+            case provisioned: return "provisioned";
+            default: throw new IllegalArgumentException("Unexpected node state '" + state + "'.");
+        }
+    }
+
+    private static String valueOf(Node.ServiceState state) {
+        switch (state) {
+            case expectedUp: return "expectedUp";
+            case allowedDown: return "allowedDown";
+            case unorchestrated: return "unorchestrated";
+            default: throw new IllegalArgumentException("Unexpected node state '" + state + "'.");
+        }
+    }
+
+    private static String valueOf(Node.ClusterType type) {
+        switch (type) {
+            case admin: return "admin";
+            case content: return "content";
+            case container: return "container";
+            default: throw new IllegalArgumentException("Unexpected node cluster type '" + type + "'.");
```

Comment:
all `valueOf` methods: default should just return "unknown" so the API doesn't break if we add types.

Reply:
Never mind, I see these are enums in the client API.